File: backend/src/compilers/examples/bernoulli_numbers.c
/**
 * Bernoulli Numbers Algorithm - Ada Lovelace's Note G (1843)
 * C Implementation with Rational Arithmetic
 *
 * This is the first computer program ever published, from Ada Lovelace's Note G
 * in her translation of Luigi Menabrea's memoir on the Analytical Engine.
 *
 * Historical Context:
 *   - Published: September 1843 in Taylor's Scientific Memoirs
 *   - Author: Augusta Ada Byron King, Countess of Lovelace
 *   - Significance: First algorithm intended for machine execution
 *   - Original: Designed for Babbage's Analytical Engine (never built)
 *
 * The Bernoulli numbers B_n are defined recursively:
 *     B_0 = 1
 *     B_1 = -1/2
 *     B_n = 0 for odd n > 1
 *     B_n = -1/(n+1) × Σ(k=0 to n-1) [C(n+1, k) × B_k] for even n > 1
 *
 * C Implementation Notes:
 *   - Uses rational number representation (numerator/denominator pairs)
 *   - GCD-based fraction simplification to prevent overflow
 *   - Memoization with fixed-size cache array
 *   - Limited to B_0 through B_30 to avoid overflow (can be extended with BigInt)
 *
 * Compilation:
 *   gcc -std=c11 -O2 -Wall -Wextra -o bernoulli bernoulli_numbers.c -lm
 *
 * Usage:
 *   ./bernoulli
 *
 * Ancient-Compute Integration:
 *   This implementation demonstrates C's low-level control while preserving
 *   the mathematical essence of Ada's original algorithm. Shows how rational
 *   arithmetic must be manually implemented in C vs. Python's fractions.Fraction.
 *
 * @file bernoulli_numbers.c
 * @author ancient-compute contributors
 * @date 2025-11-19
 * @version 1.0
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdint.h>
#include <string.h>

#define MAX_BERNOULLI_INDEX 30
#define CACHE_SIZE (MAX_BERNOULLI_INDEX + 1)

/**
 * Rational number representation.
 * Invariant: denominator > 0, gcd(numerator, denominator) = 1 (reduced form)
 */
typedef struct {
    int64_t numerator;
    int64_t denominator;
} Rational;

/**
 * Memoization cache for Bernoulli numbers.
 * cache[i] contains B_i, or {0, 0} if not yet computed.
 */
static Rational bernoulli_cache[CACHE_SIZE];
static bool cache_initialized = false;

/**
 * Compute greatest common divisor using Euclid's algorithm.
 *
 * @param a First integer
 * @param b Second integer
 * @return GCD(a, b)
 */
int64_t gcd(int64_t a, int64_t b) {
    if (a < 0) a = -a;
    if (b < 0) b = -b;

    while (b != 0) {
        int64_t temp = b;
        b = a % b;
        a = temp;
    }

    return a;
}

/**
 * Compute least common multiple.
 *
 * @param a First integer
 * @param b Second integer
 * @return LCM(a, b)
 */
int64_t lcm(int64_t a, int64_t b) {
    if (a == 0 || b == 0) return 0;
    return (a / gcd(a, b)) * b;
}

/**
 * Create a rational number in reduced form.
 *
 * @param num Numerator
 * @param den Denominator (must be non-zero)
 * @return Reduced rational number
 */
Rational make_rational(int64_t num, int64_t den) {
    if (den == 0) {
        fprintf(stderr, "Error: Division by zero\n");
        exit(1);
    }

    // Ensure denominator is positive
    if (den < 0) {
        num = -num;
        den = -den;
    }

    // Reduce to lowest terms
    if (num == 0) {
        return (Rational){0, 1};
    }

    int64_t g = gcd(num, den);
    return (Rational){num / g, den / g};
}

/**
 * Add two rational numbers.
 *
 * @param a First rational
 * @param b Second rational
 * @return a + b
 */
Rational rational_add(Rational a, Rational b) {
    // a/b + c/d = (ad + bc) / bd
    int64_t num = a.numerator * b.denominator + b.numerator * a.denominator;
    int64_t den = a.denominator * b.denominator;
    return make_rational(num, den);
}

/**
 * Multiply two rational numbers.
 *
 * @param a First rational
 * @param b Second rational
 * @return a * b
 */
Rational rational_multiply(Rational a, Rational b) {
    // (a/b) * (c/d) = (ac) / (bd)
    int64_t num = a.numerator * b.numerator;
    int64_t den = a.denominator * b.denominator;
    return make_rational(num, den);
}

/**
 * Divide two rational numbers.
 *
 * @param a First rational
 * @param b Second rational (must be non-zero)
 * @return a / b
 */
Rational rational_divide(Rational a, Rational b) {
    if (b.numerator == 0) {
        fprintf(stderr, "Error: Division by zero rational\n");
        exit(1);
    }

    // (a/b) / (c/d) = (ad) / (bc)
    int64_t num = a.numerator * b.denominator;
    int64_t den = a.denominator * b.numerator;
    return make_rational(num, den);
}

/**
 * Negate a rational number.
 *
 * @param r Rational number
 * @return -r
 */
Rational rational_negate(Rational r) {
    return make_rational(-r.numerator, r.denominator);
}

/**
 * Convert rational to double for display.
 *
 * @param r Rational number
 * @return Double approximation
 */
double rational_to_double(Rational r) {
    return (double)r.numerator / (double)r.denominator;
}

/**
 * Print a rational number in reduced form.
 *
 * @param r Rational number to print
 */
void print_rational(Rational r) {
    if (r.denominator == 1) {
        printf("%lld", (long long)r.numerator);
    } else {
        printf("%lld/%lld", (long long)r.numerator, (long long)r.denominator);
    }
}

/**
 * Compute binomial coefficient C(n, k) = n! / (k! * (n-k)!)
 *
 * Uses iterative formula to avoid factorial overflow:
 * C(n, k) = (n * (n-1) * ... * (n-k+1)) / (k * (k-1) * ... * 1)
 *
 * @param n Total items
 * @param k Items to choose
 * @return Binomial coefficient
 */
int64_t binomial_coefficient(int n, int k) {
    if (k < 0 || k > n) return 0;
    if (k == 0 || k == n) return 1;

    // Optimize: C(n, k) = C(n, n-k)
    if (k > n - k) {
        k = n - k;
    }

    int64_t result = 1;
    for (int i = 0; i < k; i++) {
        result *= (n - i);
        result /= (i + 1);
    }

    return result;
}

/**
 * Initialize the Bernoulli number cache.
 */
void init_cache(void) {
    if (cache_initialized) return;

    for (int i = 0; i <= MAX_BERNOULLI_INDEX; i++) {
        bernoulli_cache[i] = (Rational){0, 0};  // Mark as not computed
    }

    cache_initialized = true;
}

/**
 * Calculate the nth Bernoulli number using Ada Lovelace's recursive formula.
 *
 * This is the algorithm from Note G (September 1843) - the first computer program
 * ever published. Ada designed this to run on Babbage's Analytical Engine.
 *
 * Mathematical Formula:
 *     B_0 = 1
 *     B_1 = -1/2
 *     B_n = 0 for odd n > 1
 *     B_n = -1/(n+1) × Σ(k=0 to n-1) [C(n+1, k) × B_k] for even n > 1
 *
 * @param n Index of Bernoulli number (0 <= n <= MAX_BERNOULLI_INDEX)
 * @return The nth Bernoulli number as a rational
 */
Rational bernoulli_number(int n) {
    if (n < 0 || n > MAX_BERNOULLI_INDEX) {
        fprintf(stderr, "Error: Bernoulli index %d out of range [0, %d]\n",
                n, MAX_BERNOULLI_INDEX);
        exit(1);
    }

    if (!cache_initialized) {
        init_cache();
    }

    // Check cache
    if (bernoulli_cache[n].denominator != 0) {
        return bernoulli_cache[n];
    }

    Rational result;

    // Base case: B_0 = 1
    if (n == 0) {
        result = make_rational(1, 1);
    }
    // Base case: B_1 = -1/2
    else if (n == 1) {
        result = make_rational(-1, 2);
    }
    // All odd Bernoulli numbers (except B_1) are zero
    else if (n > 1 && n % 2 == 1) {
        result = make_rational(0, 1);
    }
    // Recursive formula for even n > 1
    else {
        // B_n = -1/(n+1) × Σ(k=0 to n-1) [C(n+1, k) × B_k]
        Rational summation = make_rational(0, 1);

        for (int k = 0; k < n; k++) {
            int64_t binom_coeff = binomial_coefficient(n + 1, k);
            Rational B_k = bernoulli_number(k);  // Recursive call

            Rational term = rational_multiply(
                make_rational(binom_coeff, 1),
                B_k
            );

            summation = rational_add(summation, term);
        }

        // B_n = -summation / (n + 1)
        Rational divisor = make_rational(n + 1, 1);
        result = rational_negate(rational_divide(summation, divisor));
    }

    // Cache result
    bernoulli_cache[n] = result;

    return result;
}

/**
 * Print a table of Bernoulli numbers.
 *
 * @param max_n Maximum index to display
 */
void print_bernoulli_table(int max_n) {
    printf("n    B_n (exact)              B_n (decimal)\n");
    printf("======================================================================\n");

    for (int i = 0; i <= max_n; i++) {
        Rational b_n = bernoulli_number(i);

        // Only print non-zero values
        if (b_n.numerator != 0) {
            printf("%-4d ", i);
            print_rational(b_n);
            printf("%*s", (int)(24 - strlen("")), "");  // Padding
            printf("%.15f\n", rational_to_double(b_n));
        }
    }
}

/**
 * Demonstrate Ada Lovelace's Note G example.
 *
 * Ada computed B_8 in her 1843 publication as the demonstration
 * of the Analytical Engine's capabilities.
 */
void ada_lovelace_note_g_example(void) {
    printf("Ada Lovelace's Note G - Bernoulli Numbers Calculation\n");
    printf("======================================================================\n\n");

    printf("Original Context:\n");
    printf("  Published: September 1843\n");
    printf("  Machine: Babbage's Analytical Engine (never built)\n");
    printf("  Variables: 9 columns (V1-V9)\n");
    printf("  Operations: 25 steps (Load, Add, Subtract, Multiply, Divide)\n\n");

    // Compute B_8 using Ada's formula
    int n = 8;
    printf("Computing B_%d using Ada's recursive formula...\n\n", n);

    Rational b_8 = bernoulli_number(n);

    printf("Result: B_%d = ", n);
    print_rational(b_8);
    printf("\n");
    printf("Decimal: %.15f\n\n", rational_to_double(b_8));

    printf("Verification - First 10 Bernoulli numbers:\n");
    print_bernoulli_table(10);
}

/**
 * Test suite for Bernoulli number implementation.
 */
void run_tests(void) {
    printf("\n=== Running Bernoulli Number Tests ===\n\n");

    int passed = 0, failed = 0;

    // Test 1: B_0 = 1
    Rational b0 = bernoulli_number(0);
    if (b0.numerator == 1 && b0.denominator == 1) {
        printf("✓ Test 1 passed: B_0 = 1\n");
        passed++;
    } else {
        printf("✗ Test 1 failed: B_0 != 1\n");
        failed++;
    }

    // Test 2: B_1 = -1/2
    Rational b1 = bernoulli_number(1);
    if (b1.numerator == -1 && b1.denominator == 2) {
        printf("✓ Test 2 passed: B_1 = -1/2\n");
        passed++;
    } else {
        printf("✗ Test 2 failed: B_1 != -1/2\n");
        failed++;
    }

    // Test 3: B_2 = 1/6
    Rational b2 = bernoulli_number(2);
    if (b2.numerator == 1 && b2.denominator == 6) {
        printf("✓ Test 3 passed: B_2 = 1/6\n");
        passed++;
    } else {
        printf("✗ Test 3 failed: B_2 != 1/6\n");
        failed++;
    }

    // Test 4: B_3 = 0 (odd number theorem)
    Rational b3 = bernoulli_number(3);
    if (b3.numerator == 0) {
        printf("✓ Test 4 passed: B_3 = 0 (odd number theorem)\n");
        passed++;
    } else {
        printf("✗ Test 4 failed: B_3 != 0\n");
        failed++;
    }

    // Test 5: B_4 = -1/30
    Rational b4 = bernoulli_number(4);
    if (b4.numerator == -1 && b4.denominator == 30) {
        printf("✓ Test 5 passed: B_4 = -1/30\n");
        passed++;
    } else {
        printf("✗ Test 5 failed: B_4 != -1/30\n");
        failed++;
    }

    // Test 6: B_8 = -1/30 (Ada's example)
    Rational b8 = bernoulli_number(8);
    if (b8.numerator == -1 && b8.denominator == 30) {
        printf("✓ Test 6 passed: B_8 = -1/30 (Ada's example)\n");
        passed++;
    } else {
        printf("✗ Test 6 failed: B_8 != -1/30\n");
        failed++;
    }

    // Test 7: B_10 = 5/66
    Rational b10 = bernoulli_number(10);
    if (b10.numerator == 5 && b10.denominator == 66) {
        printf("✓ Test 7 passed: B_10 = 5/66\n");
        passed++;
    } else {
        printf("✗ Test 7 failed: B_10 != 5/66\n");
        failed++;
    }

    // Test 8: B_12 = -691/2730
    Rational b12 = bernoulli_number(12);
    if (b12.numerator == -691 && b12.denominator == 2730) {
        printf("✓ Test 8 passed: B_12 = -691/2730\n");
        passed++;
    } else {
        printf("✗ Test 8 failed: B_12 != -691/2730\n");
        failed++;
    }

    printf("\n=== Test Results ===\n");
    printf("Passed: %d/%d\n", passed, passed + failed);
    printf("Failed: %d/%d\n", failed, passed + failed);

    if (failed == 0) {
        printf("\n✓ All tests passed!\n");
    }
}

/**
 * Main entry point.
 */
int main(void) {
    // Run Ada's historical example
    ada_lovelace_note_g_example();

    // Run test suite
    run_tests();

    return 0;
}

File: backend/src/compilers/examples/BernoulliNumbers.java
/**
 * Bernoulli Numbers Algorithm - Ada Lovelace's Note G (1843)
 * Java Implementation with BigInteger Rational Arithmetic
 *
 * This is the first computer program ever published, from Ada Lovelace's Note G
 * in her translation of Luigi Menabrea's memoir on the Analytical Engine.
 *
 * Historical Context:
 *   - Published: September 1843 in Taylor's Scientific Memoirs
 *   - Author: Augusta Ada Byron King, Countess of Lovelace
 *   - Significance: First algorithm intended for machine execution
 *   - Original: Designed for Babbage's Analytical Engine (never built)
 *
 * The Bernoulli numbers B_n are defined recursively:
 *     B_0 = 1
 *     B_1 = -1/2
 *     B_n = 0 for odd n > 1
 *     B_n = -1/(n+1) × Σ(k=0 to n-1) [C(n+1, k) × B_k] for even n > 1
 *
 * Java Implementation Features:
 *   - Object-oriented design with Rational class
 *   - BigInteger for arbitrary precision arithmetic
 *   - HashMap-based memoization for efficiency
 *   - Comprehensive test suite with JUnit-style assertions
 *   - Demonstrates OOP paradigm applied to mathematical computation
 *
 * Compilation:
 *   javac BernoulliNumbers.java
 *
 * Usage:
 *   java BernoulliNumbers
 *
 * Ancient-Compute Integration:
 *   This implementation demonstrates Java's object-oriented approach to
 *   mathematical computation, contrasting with functional (Haskell) and
 *   procedural (C) approaches while preserving Ada's algorithm logic.
 *
 * @author ancient-compute contributors
 * @version 1.0
 * @since 2025-11-19
 */

import java.math.BigInteger;
import java.util.HashMap;
import java.util.Map;

/**
 * Immutable rational number class using BigInteger for numerator and denominator.
 * Maintains invariant: denominator > 0, gcd(numerator, denominator) = 1.
 */
class Rational {
    private final BigInteger numerator;
    private final BigInteger denominator;

    /**
     * Constructs a rational number in reduced form.
     *
     * @param num Numerator
     * @param den Denominator (must be non-zero)
     * @throws ArithmeticException if denominator is zero
     */
    public Rational(BigInteger num, BigInteger den) {
        if (den.equals(BigInteger.ZERO)) {
            throw new ArithmeticException("Denominator cannot be zero");
        }

        // Ensure denominator is positive
        if (den.compareTo(BigInteger.ZERO) < 0) {
            num = num.negate();
            den = den.negate();
        }

        // Reduce to lowest terms
        if (num.equals(BigInteger.ZERO)) {
            this.numerator = BigInteger.ZERO;
            this.denominator = BigInteger.ONE;
        } else {
            BigInteger gcd = num.gcd(den);
            this.numerator = num.divide(gcd);
            this.denominator = den.divide(gcd);
        }
    }

    /**
     * Convenience constructor for integer values.
     *
     * @param num Numerator (as long)
     * @param den Denominator (as long)
     */
    public Rational(long num, long den) {
        this(BigInteger.valueOf(num), BigInteger.valueOf(den));
    }

    /**
     * Convenience constructor for integer rational (denominator = 1).
     *
     * @param num Numerator
     */
    public Rational(BigInteger num) {
        this(num, BigInteger.ONE);
    }

    /**
     * Get numerator.
     *
     * @return Numerator of this rational number
     */
    public BigInteger getNumerator() {
        return numerator;
    }

    /**
     * Get denominator.
     *
     * @return Denominator of this rational number
     */
    public BigInteger getDenominator() {
        return denominator;
    }

    /**
     * Add two rational numbers.
     *
     * @param other Rational number to add
     * @return Sum of this and other
     */
    public Rational add(Rational other) {
        // a/b + c/d = (ad + bc) / bd
        BigInteger num = this.numerator.multiply(other.denominator)
                        .add(other.numerator.multiply(this.denominator));
        BigInteger den = this.denominator.multiply(other.denominator);
        return new Rational(num, den);
    }

    /**
     * Multiply two rational numbers.
     *
     * @param other Rational number to multiply
     * @return Product of this and other
     */
    public Rational multiply(Rational other) {
        // (a/b) * (c/d) = (ac) / (bd)
        BigInteger num = this.numerator.multiply(other.numerator);
        BigInteger den = this.denominator.multiply(other.denominator);
        return new Rational(num, den);
    }

    /**
     * Divide two rational numbers.
     *
     * @param other Rational number to divide by (must be non-zero)
     * @return Quotient of this and other
     * @throws ArithmeticException if other is zero
     */
    public Rational divide(Rational other) {
        if (other.numerator.equals(BigInteger.ZERO)) {
            throw new ArithmeticException("Division by zero rational");
        }

        // (a/b) / (c/d) = (ad) / (bc)
        BigInteger num = this.numerator.multiply(other.denominator);
        BigInteger den = this.denominator.multiply(other.numerator);
        return new Rational(num, den);
    }

    /**
     * Negate this rational number.
     *
     * @return Negation of this rational
     */
    public Rational negate() {
        return new Rational(this.numerator.negate(), this.denominator);
    }

    /**
     * Convert to double for numerical applications.
     *
     * @return Double approximation of this rational
     */
    public double toDouble() {
        return numerator.doubleValue() / denominator.doubleValue();
    }

    /**
     * String representation in reduced form.
     *
     * @return String representation "num/den" or "num" if denominator is 1
     */
    @Override
    public String toString() {
        if (denominator.equals(BigInteger.ONE)) {
            return numerator.toString();
        } else {
            return numerator + "/" + denominator;
        }
    }

    /**
     * Equality comparison.
     *
     * @param obj Object to compare with
     * @return true if obj is a Rational equal to this
     */
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof Rational)) return false;

        Rational other = (Rational) obj;
        return this.numerator.equals(other.numerator) &&
               this.denominator.equals(other.denominator);
    }

    /**
     * Hash code for use in collections.
     *
     * @return Hash code value
     */
    @Override
    public int hashCode() {
        return numerator.hashCode() * 31 + denominator.hashCode();
    }
}

/**
 * Main class implementing Ada Lovelace's Bernoulli Numbers Algorithm.
 */
public class BernoulliNumbers {
    // Memoization cache for computed Bernoulli numbers
    private static final Map<Integer, Rational> cache = new HashMap<>();

    /**
     * Compute binomial coefficient C(n, k) = n! / (k! * (n-k)!)
     *
     * Uses iterative formula to avoid factorial overflow:
     * C(n, k) = (n * (n-1) * ... * (n-k+1)) / (k * (k-1) * ... * 1)
     *
     * @param n Total items
     * @param k Items to choose
     * @return Binomial coefficient
     */
    public static BigInteger binomialCoefficient(int n, int k) {
        if (k < 0 || k > n) return BigInteger.ZERO;
        if (k == 0 || k == n) return BigInteger.ONE;

        // Optimization: C(n, k) = C(n, n-k)
        if (k > n - k) {
            k = n - k;
        }

        BigInteger result = BigInteger.ONE;
        for (int i = 0; i < k; i++) {
            result = result.multiply(BigInteger.valueOf(n - i));
            result = result.divide(BigInteger.valueOf(i + 1));
        }

        return result;
    }

    /**
     * Calculate the nth Bernoulli number using Ada Lovelace's recursive formula.
     *
     * This is the algorithm from Note G (September 1843) - the first computer program
     * ever published. Ada designed this to run on Babbage's Analytical Engine.
     *
     * Mathematical Formula:
     *     B_0 = 1
     *     B_1 = -1/2
     *     B_n = 0 for odd n > 1
     *     B_n = -1/(n+1) × Σ(k=0 to n-1) [C(n+1, k) × B_k] for even n > 1
     *
     * @param n Index of Bernoulli number (n >= 0)
     * @return The nth Bernoulli number as a Rational
     * @throws IllegalArgumentException if n < 0
     */
    public static Rational bernoulliNumber(int n) {
        if (n < 0) {
            throw new IllegalArgumentException("Bernoulli number index must be non-negative, got " + n);
        }

        // Check cache first
        if (cache.containsKey(n)) {
            return cache.get(n);
        }

        Rational result;

        // Base case: B_0 = 1
        if (n == 0) {
            result = new Rational(1, 1);
        }
        // Base case: B_1 = -1/2
        else if (n == 1) {
            result = new Rational(-1, 2);
        }
        // All odd Bernoulli numbers (except B_1) are zero
        else if (n % 2 == 1) {
            result = new Rational(0, 1);
        }
        // Recursive formula for even n > 1
        else {
            // B_n = -1/(n+1) × Σ(k=0 to n-1) [C(n+1, k) × B_k]
            Rational summation = new Rational(0, 1);

            for (int k = 0; k < n; k++) {
                BigInteger binomCoeff = binomialCoefficient(n + 1, k);
                Rational Bk = bernoulliNumber(k);  // Recursive call

                Rational term = new Rational(binomCoeff).multiply(Bk);
                summation = summation.add(term);
            }

            // B_n = -summation / (n + 1)
            Rational divisor = new Rational(n + 1, 1);
            result = summation.divide(divisor).negate();
        }

        // Cache result
        cache.put(n, result);

        return result;
    }

    /**
     * Generate a sequence of Bernoulli numbers from B_0 to B_max.
     *
     * @param maxN Maximum index to compute
     * @return Array of Bernoulli numbers
     */
    public static Rational[] bernoulliSequence(int maxN) {
        if (maxN < 0) {
            throw new IllegalArgumentException("max_n must be non-negative, got " + maxN);
        }

        Rational[] sequence = new Rational[maxN + 1];
        for (int i = 0; i <= maxN; i++) {
            sequence[i] = bernoulliNumber(i);
        }

        return sequence;
    }

    /**
     * Print a formatted table of Bernoulli numbers.
     *
     * @param maxN Maximum index to display
     */
    public static void printBernoulliTable(int maxN) {
        System.out.println("n    B_n (exact)              B_n (decimal)");
        System.out.println("======================================================================");

        for (int i = 0; i <= maxN; i++) {
            Rational bn = bernoulliNumber(i);

            // Only print non-zero values
            if (!bn.getNumerator().equals(BigInteger.ZERO)) {
                System.out.printf("%-4d %-24s %.15f%n",
                    i, bn.toString(), bn.toDouble());
            }
        }
    }

    /**
     * Demonstrate Ada Lovelace's Note G example.
     *
     * Ada computed B_8 in her 1843 publication as the demonstration
     * of the Analytical Engine's capabilities.
     */
    public static void adaLovelaceNoteGExample() {
        System.out.println("Ada Lovelace's Note G - Bernoulli Numbers Calculation");
        System.out.println("======================================================================");
        System.out.println();
        System.out.println("Original Context:");
        System.out.println("  Published: September 1843");
        System.out.println("  Machine: Babbage's Analytical Engine (never built)");
        System.out.println("  Variables: 9 columns (V1-V9)");
        System.out.println("  Operations: 25 steps (Load, Add, Subtract, Multiply, Divide)");
        System.out.println();

        int n = 8;
        System.out.println("Computing B_" + n + " using Ada's recursive formula...");
        System.out.println();

        Rational b8 = bernoulliNumber(n);

        System.out.println("Result: B_" + n + " = " + b8);
        System.out.printf("Decimal: %.15f%n", b8.toDouble());
        System.out.println();

        System.out.println("Verification - First 10 Bernoulli numbers:");
        printBernoulliTable(10);
    }

    /**
     * Test suite for Bernoulli number implementation.
     */
    public static void runTests() {
        System.out.println();
        System.out.println("=== Running Bernoulli Number Tests ===");
        System.out.println();

        int passed = 0;
        int failed = 0;

        // Test 1: B_0 = 1
        if (bernoulliNumber(0).equals(new Rational(1, 1))) {
            System.out.println("✓ Test 1 passed: B_0 = 1");
            passed++;
        } else {
            System.out.println("✗ Test 1 failed: B_0 != 1");
            failed++;
        }

        // Test 2: B_1 = -1/2
        if (bernoulliNumber(1).equals(new Rational(-1, 2))) {
            System.out.println("✓ Test 2 passed: B_1 = -1/2");
            passed++;
        } else {
            System.out.println("✗ Test 2 failed: B_1 != -1/2");
            failed++;
        }

        // Test 3: B_2 = 1/6
        if (bernoulliNumber(2).equals(new Rational(1, 6))) {
            System.out.println("✓ Test 3 passed: B_2 = 1/6");
            passed++;
        } else {
            System.out.println("✗ Test 3 failed: B_2 != 1/6");
            failed++;
        }

        // Test 4: B_3 = 0 (odd number theorem)
        if (bernoulliNumber(3).equals(new Rational(0, 1))) {
            System.out.println("✓ Test 4 passed: B_3 = 0 (odd number theorem)");
            passed++;
        } else {
            System.out.println("✗ Test 4 failed: B_3 != 0");
            failed++;
        }

        // Test 5: B_4 = -1/30
        if (bernoulliNumber(4).equals(new Rational(-1, 30))) {
            System.out.println("✓ Test 5 passed: B_4 = -1/30");
            passed++;
        } else {
            System.out.println("✗ Test 5 failed: B_4 != -1/30");
            failed++;
        }

        // Test 6: B_8 = -1/30 (Ada's example)
        if (bernoulliNumber(8).equals(new Rational(-1, 30))) {
            System.out.println("✓ Test 6 passed: B_8 = -1/30 (Ada's example)");
            passed++;
        } else {
            System.out.println("✗ Test 6 failed: B_8 != -1/30");
            failed++;
        }

        // Test 7: B_10 = 5/66
        if (bernoulliNumber(10).equals(new Rational(5, 66))) {
            System.out.println("✓ Test 7 passed: B_10 = 5/66");
            passed++;
        } else {
            System.out.println("✗ Test 7 failed: B_10 != 5/66");
            failed++;
        }

        // Test 8: B_12 = -691/2730
        if (bernoulliNumber(12).equals(new Rational(-691, 2730))) {
            System.out.println("✓ Test 8 passed: B_12 = -691/2730");
            passed++;
        } else {
            System.out.println("✗ Test 8 failed: B_12 != -691/2730");
            failed++;
        }

        // Test 9: B_20 = -174611/330
        if (bernoulliNumber(20).equals(new Rational(
                new BigInteger("-174611"),
                new BigInteger("330")))) {
            System.out.println("✓ Test 9 passed: B_20 = -174611/330");
            passed++;
        } else {
            System.out.println("✗ Test 9 failed: B_20 != -174611/330");
            failed++;
        }

        // Test 10: Sequence length
        if (bernoulliSequence(10).length == 11) {
            System.out.println("✓ Test 10 passed: Sequence length");
            passed++;
        } else {
            System.out.println("✗ Test 10 failed: Sequence length");
            failed++;
        }

        System.out.println();
        System.out.println("=== Test Results ===");
        System.out.printf("Passed: %d/%d%n", passed, passed + failed);
        System.out.printf("Failed: %d/%d%n", failed, passed + failed);
        System.out.println();

        if (failed == 0) {
            System.out.println("✓ All tests passed!");
        } else {
            System.out.println("✗ Some tests failed");
        }
    }

    /**
     * Main entry point.
     *
     * @param args Command line arguments (unused)
     */
    public static void main(String[] args) {
        // Run Ada's historical example
        adaLovelaceNoteGExample();

        // Run test suite
        runTests();
    }
}
